### PR TITLE
Refactor add api to generate resources based on AddApiRequest object

### PR DESCRIPTION
**Pull request by edwardfoyle**
Refactored add api flow to generate the AddApiRequest object and then do resource generation based on that.

I haven't yet wired up the new resource generation to a headless entry point. Also haven't fully tested these changes, but going ahead and putting up a PR to get feedback.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/graphql-transformer-core/src/collectDirectives.ts [hunk 1]
```diff
@@ -47,7 +47,7 @@ export function collectDirectives(sdl: string): DirectiveNode[] {
   return directives;
 }
 
-export function collectDirectivesByTypeNames(sdl: string): Object {
+export function collectDirectivesByTypeNames(sdl: string) {
```
Comment (kaustavghosh06) [inline packages/graphql-transformer-core/src/collectDirectives.ts:50]: Should this be part of this change?

File: packages/amplify-util-headless-input/package.json [hunk 2]
```diff
@@ -5,7 +5,7 @@
   "main": "lib/index.js",
   "scripts": {
     "build": "tsc",
-    "clean": "rimraf ./lib && rm tsconfig.tsbuildinfo",
+    "clean": "rimraf lib tsconfig.tsbuildinfo",
```
Comment (kaustavghosh06) [inline packages/amplify-util-headless-input/package.json:8]: Is rimraf a valid command in Windows?

File: packages/amplify-headless-interface/src/interface/api/add.ts [hunk 3]
```diff
@@ -18,18 +18,43 @@ export interface AppSyncServiceConfiguration {
...
+  defaultResolutionStrategy?: ResolutionStrategy;
+  perModelResolutionStrategy?: PerModelResolutionstrategy[];
 }
 
+export interface PerModelResolutionstrategy {
+  resolutionStrategy: ResolutionStrategy;
+  entityName: string;
+}
+
+export interface PredefinedResolutionStrategy {
```
Comment (kaustavghosh06) [inline packages/amplify-headless-interface/src/interface/api/add.ts:34]: Why have we separated Lambda resolution strategy out from the other resolution strategies?

File: packages/amplify-category-api/resources/awscloudformation/graphql-schemas/basic-schema.graphql.ejs [hunk 4]
```diff
@@ -1,7 +0,0 @@
-type <%= props.typeName %> @model {
```
Comment (kaustavghosh06) [inline packages/amplify-category-api/resources/awscloudformation/graphql-schemas/basic-schema.graphql.ejs:1]: Have we removed the base type question? We should probably have an API review for it and separate that out in another PR since it's an independent concern and not tied to this functionality. Thoughts?

File: packages/amplify-category-api/src/__tests__/apigw-walkthrough.test.js [hunk 5]
```diff
@@ -1,6 +0,0 @@
-const { getIAMPolicies } = require('../provider-utils/awscloudformation/service-walkthroughs/apigw-walkthrough');
-
-test('getIAMPolicies', () => {
-  output = getIAMPolicies('resourceName', ['read']);
-  expect(output.attributes).toStrictEqual(['ApiName', 'ApiId']);
```
Comment (kaustavghosh06) [inline packages/amplify-category-api/src/__tests__/apigw-walkthrough.test.js:5]: Did we move this test to some place else now?

File: packages/amplify-category-api/src/provider-utils/awscloudformation/CfnApiArtifactHandler.ts [hunk 6]
```diff
@@ -0,0 +1,176 @@
+import { ApiArtifactHandler } from '../ApiArtifactHandler';
```
Comment (kaustavghosh06) [inline packages/amplify-category-api/src/provider-utils/awscloudformation/CfnApiArtifactHandler.ts:1]: I think we should standardize naming of our files. We're using kebab-case for majority of our existing file names.

File: packages/amplify-category-api/src/provider-utils/awscloudformation/awsConstants.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+import path from 'path';
```
Comment (kaustavghosh06) [inline packages/amplify-category-api/src/provider-utils/awscloudformation/awsConstants.ts:1]: I think we have a constants file in src/ - can't we re-use that?

File: packages/amplify-category-api/src/provider-utils/awscloudformation/service-walkthroughs/appSync-walkthrough.ts [hunk 8]
```diff
@@ -1,28 +1,23 @@
...
   readTransformerConfiguration,
   writeTransformerConfiguration,
-  TRANSFORM_CURRENT_VERSION,
-} = TransformPackage;
+  readProjectConfiguration,
+} from 'graphql-transformer-core';
+import { category } from '../../../categoryConstants';
+import { writeResolverConfig } from '../CfnApiArtifactHandler';
+
+const serviceName = 'AppSync';
```
Comment (kaustavghosh06) [inline packages/amplify-category-api/src/provider-utils/awscloudformation/service-walkthroughs/appSync-walkthrough.ts:18]: Does this change the serviceName we store in amplify-meta? If so - this change might break existing projects. We use the service-name in a variety of packages including code-gen.

File: packages/amplify-category-api/src/provider-utils/awscloudformation/utils/authConfigToAppSyncAuthTypeBiDiMapper.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+  return _.get(authConfigToAppSyncAuthTypeMap, authConfig.authenticationType, () => undefined)(authConfig);
+};
+
+export const appSyncAuthTypeToAuthConfig = (authType?: AppSyncAuthType) => {
+  if (!authType) return undefined;
+  return _.get(appSyncAuthTypeToAuthConfigMap, authType.mode, () => undefined)(authType);
+};
+
+const authConfigToAppSyncAuthTypeMap: Record<string, (oldAuthObj: any) => AppSyncAuthType> = {
+  API_KEY: oldAuthObj => ({
```
Comment (kaustavghosh06) [inline packages/amplify-category-api/src/provider-utils/awscloudformation/utils/authConfigToAppSyncAuthTypeBiDiMapper.ts:19]: What do you refer to as oldAuthObj? Legacy?

File: packages/amplify-category-api/src/provider-utils/awscloudformation/utils/authConfigToAppSyncAuthTypeBiDiMapper.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
+import {
+  AppSyncAuthType,
+  AppSyncAPIKeyAuthType,
+  AppSyncCognitoUserPoolsAuthType,
+  AppSyncOpenIDConnectAuthType,
+} from 'amplify-headless-interface';
+import _ from 'lodash';
+
+export const authConfigToAppSyncAuthType = (authConfig: any = {}): AppSyncAuthType => {
```
Comment (kaustavghosh06) [inline packages/amplify-category-api/src/provider-utils/awscloudformation/utils/authConfigToAppSyncAuthTypeBiDiMapper.ts:9]: Could you provide some comment on top these two functions as to what's the purpose of these?

File: packages/graphql-transformer-core/src/collectDirectives.ts (hunk 1 above)
Comment (edwardfoyle) [reply packages/graphql-transformer-core/src/collectDirectives.ts:50]: Yeah we use this method in the api category and removing this allows TS to infer the type from the return value

File: packages/amplify-util-headless-input/package.json (hunk 2 above)
Comment (edwardfoyle) [reply packages/amplify-util-headless-input/package.json:8]: It's a node package. rm is the command that doesn't work on windows

File: packages/amplify-category-api/src/provider-utils/awscloudformation/utils/serviceWalkthroughResultToAddApiRequest.ts [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+// Long-term, the service walkthrough should be refactored to directly return an object conforming to the interface
+export const serviceWalkthroughResultToAddApiRequest = (result): AddApiRequest => ({
+  version: 1,
+  serviceConfiguration: {
+    serviceName: 'AppSync',
+    apiName: result.answers.apiName,
+    transformSchema: result.schemaContent,
+    defaultAuthType: authConfigToAppSyncAuthType(result.output.authConfig.defaultAuthentication),
+    additionalAuthTypes: (result.output.authConfig.additionalAuthenticationProviders || []).map(authConfigToAppSyncAuthType),
+    conflictResolution: resolverConfigToConflictResolution(result.resolverConfig),
```
[the remark this answers is not in the document]
Comment (edwardfoyle) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/utils/serviceWalkthroughResultToAddApiRequest.ts:16]: if it's undefined, it will get transformed to an empty object

File: packages/amplify-category-api/src/provider-utils/awscloudformation/utils/restApiPathUtils.ts [hunk 12]
```diff
@@ -0,0 +1,37 @@
+export const validatePathName = (name: string, paths: { name: string }[]) => {
```
[the remark this answers is not in the document]
Comment (edwardfoyle) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/utils/restApiPathUtils.ts:1]: they used to be in the walkthrough file but I put them in this file so I could remove some jank testing we had for them (the tests are still there, they just aren't jank anymore). But I didn't change any functionality

File: packages/amplify-category-api/src/provider-utils/awscloudformation/utils/dynamicImports.ts [hunk 13]
```diff
@@ -0,0 +1,4 @@
+export const serviceMetadataFor = async service => (await import('../../supported-services')).supportedServices[service];
```
[the remark this answers is not in the document]
Comment (edwardfoyle) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/utils/dynamicImports.ts:1]: I put them here so we have a central location where we can see all of our dynamic imports rather than having them scattered all over the place

File: packages/amplify-category-api/src/provider-utils/awscloudformation/service-walkthroughs/apigw-walkthrough.ts [hunk 14]
```diff
@@ -73,8 +74,8 @@ async function updateWalkthrough(context, defaultValuesFilename) {
   }
 
   const projectBackendDirPath = context.amplify.pathManager.getBackendDirPath();
-  const resourceDirPath = pathLib.join(projectBackendDirPath, category, updateApi.resourceName);
-  const parametersFilePath = pathLib.join(resourceDirPath, parametersFileName);
+  const resourceDirPath = path.join(projectBackendDirPath, category, updateApi.resourceName as string);
```
[the remark this answers is not in the document]
Comment (edwardfoyle) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/service-walkthroughs/apigw-walkthrough.ts:77]: there's no other assignment to path in this closure. I'll run through the rest walkthrough manually to make sure there's no problems though

File: packages/amplify-category-api/resources/awscloudformation/graphql-schemas/basic-schema.graphql.ejs (hunk 4 above)
Comment (edwardfoyle) [reply packages/amplify-category-api/resources/awscloudformation/graphql-schemas/basic-schema.graphql.ejs:1]: The problem with pulling it into a separate PR is that it makes this one much more complicated. It's the only schema that requires a value from the walkthrough in order to copy which means to separate the walkthrough and the file i/o would require doing the replacement to a temp file, then copy that to the project and delete the temp file which seems like way more effort than it's worth

File: packages/amplify-category-api/src/__tests__/apigw-walkthrough.test.js (hunk 5 above)
Comment (edwardfoyle) [reply packages/amplify-category-api/src/__tests__/apigw-walkthrough.test.js:5]: yes, I didn't remove any tests, I just moved them into a folder structure within __tests__ that mirrors their location in src

File: packages/amplify-category-api/src/provider-utils/awscloudformation/CfnApiArtifactHandler.ts (hunk 6 above)
Comment (edwardfoyle) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/CfnApiArtifactHandler.ts:1]: I don't have super strong opinions here other than consistency, but it seems to be customary for files that define classes / interfaces to have the class name as the file name which is what I did here

File: packages/amplify-category-api/src/provider-utils/awscloudformation/awsConstants.ts (hunk 7 above)
Comment (edwardfoyle) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/awsConstants.ts:1]: I made this file because these constants are specific to the aws cfn provider so they don't really apply across the whole category (although there's only one provider)

File: packages/amplify-category-api/src/provider-utils/awscloudformation/service-walkthroughs/appSync-walkthrough.ts (hunk 8 above)
Comment (edwardfoyle) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/service-walkthroughs/appSync-walkthrough.ts:18]: No, this is the existing service name and I changed the service name enum in the spec to match this

File: packages/amplify-category-api/src/provider-utils/awscloudformation/utils/authConfigToAppSyncAuthTypeBiDiMapper.ts (hunk 9 above)
Comment (edwardfoyle) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/utils/authConfigToAppSyncAuthTypeBiDiMapper.ts:19]: it's the authConfig object that comes out of the existing walkthrough. I'll change this variable to just "authConfig"

File: packages/amplify-category-api/src/provider-utils/awscloudformation/utils/authConfigToAppSyncAuthTypeBiDiMapper.ts (hunk 10 above)
Comment (edwardfoyle) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/utils/authConfigToAppSyncAuthTypeBiDiMapper.ts:9]: done

File: packages/amplify-category-api/src/provider-utils/awscloudformation/utils/serviceWalkthroughResultToAddApiRequest.ts [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+
+// Temporary conversion function between the existing output of the appSync service walkthrough and the new AddApiRequest interface
+// Long-term, the service walkthrough should be refactored to directly return an object conforming to the interface
+export const serviceWalkthroughResultToAddApiRequest = (result): AddApiRequest => ({
+  version: 1,
+  serviceConfiguration: {
+    serviceName: 'AppSync',
+    apiName: result.answers.apiName,
+    transformSchema: result.schemaContent,
+    defaultAuthType: authConfigToAppSyncAuthType(result.output.authConfig.defaultAuthentication),
```
[the remark this answers is not in the document]
Comment (edwardfoyle) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/utils/serviceWalkthroughResultToAddApiRequest.ts:14]: I don't think it's necessary. This translation works and if we make more changes to the walkthrough, we should just refactor it to return the AddApiRequest directly so this translation layer isn't even necessary

File: packages/amplify-category-api/src/provider-utils/awscloudformation/service-walkthroughs/appSync-walkthrough.ts (hunk 8 above)
Comment (kaustavghosh06) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/service-walkthroughs/appSync-walkthrough.ts:12]: Does this change the service name which we store in the amplify-meta though?

Comment (edwardfoyle) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/service-walkthroughs/appSync-walkthrough.ts:12]: no, "AppSync" is what we currently store in amplify-meta

File: packages/amplify-category-api/src/provider-utils/awscloudformation/service-walkthroughs/apigw-walkthrough.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (kaustavghosh06) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/service-walkthroughs/apigw-walkthrough.ts:77]: Yeah, please check on this once and then resolve the conversation.

File: packages/amplify-category-api/resources/awscloudformation/graphql-schemas/basic-schema.graphql.ejs (hunk 4 above)
Comment (kaustavghosh06) [reply packages/amplify-category-api/resources/awscloudformation/graphql-schemas/basic-schema.graphql.ejs:1]: Okay, let's do a short API review as well as suggested by Richard before we merge this to master.

File: packages/amplify-category-api/src/provider-utils/awscloudformation/CfnApiArtifactHandler.ts (hunk 6 above)
Comment (kaustavghosh06) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/CfnApiArtifactHandler.ts]: Discussed offline and file-names have been changed to kebab-case.

File: packages/amplify-category-api/src/provider-utils/awscloudformation/service-walkthroughs/apigw-walkthrough.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (edwardfoyle) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/service-walkthroughs/apigw-walkthrough.ts:77]: just went through the walkthrough with no problems
